New SCAN relaxation workflow

rkingsbury opened the pull request:
## Summary

This commit introduces a new workflow for structure optimization using the SCAN metaGGA functional. It depends on a new VaspInputSet (`MPScanRelaxSet`) that was recently added to pymatgen (see [PR](https://github.com/materialsproject/pymatgen/pull/1708)). (Note: tests will fail until the latest pymatgen is pulled into atomate's CI).

Consistent with my understanding of the design intent of atomate, I've taken steps to invoke custodian **only** for error handling via the `use_custodian` powerup, and to put all other tasks (job construction, file moving/renaming, gzipping output, etc.) into atomate. This has required a few changes to the `glue_tasks` methods. 

The workflow is able to run without custodian if desired (by disabling the powerup), and all the expected outputs (.relax1 and .relax2 suffixes, DB insertion, etc.) will still be generated.

rkingsbury commented:
@utf I'm interested in your feedback on this PR, and input on how to address the TODO items. Thanks!

mkhorton commented:
Could you comment a little on `CopyFilesFromCalcLoc` changes etc?

rkingsbury commented:
> Could you comment a little on `CopyFilesFromCalcLoc` changes etc?

Sure @mkhorton . Because I'm moving the file operations (backup, rename, gzip, etc.) into atomate, I had to add some new capabilities to the glue_tasks. Specifically:

- Added  glob pattern support and the ability to exclude files to `CopyFilesFromCalcLoc` (similar to what already existed in `DeleteFiles`
- Added env_chk support to `CopyFiles`. This was necessary to facilitate copying over a pre-compiled vdw kernel from a central location on Cori
- Added a new GzipDir glue_task that .gzips everything in a directory

rkingsbury commented:
@mkhorton @computron @utf  In order for tests to pass, the CI will need access to the latest-and-greatest pymatgen code (containing `MPScanRelaxSet`. Can you please update the CI?

mkhorton commented:
@rkingsbury the latest-and-greatest pymatgen doesn't have MPScanRelaxSet yet, I'd have to push a release

rkingsbury commented:
> @rkingsbury the latest-and-greatest pymatgen doesn't have MPScanRelaxSet yet, I'd have to push a release

Ah, right.

mkhorton commented:
There was a discussion of switching the CI to pull from pymatgen master, I think it was causing similar issues for @samblau's development, but for now we're still pinned.

bocklund commented:
@mkhorton the test matrix could do both: one test in the matrix against release pymatgen and another against the master branch

utf commented:
Can you pull in the latest changes from master? I've updated the pymatgen requirement to the latest version so this should fix the failing tests.

utf commented on atomate/vasp/workflows/presets/scan.py:
```diff
@@ -1,50 +1,57 @@
-            "job_type": "metagga_opt_run",
-            "vasp_cmd": vasp_cmd
-        })
+        wf, custodian_params={"job_type": "normal_no_backup", "gzip_output": False}
```
Inline with the rest of atomate, I think using Custodian in the Firework itself should be the default. And users can apply the `remove_custodian` firework if they want to.

rkingsbury commented:
defaulting to `RunVaspCustodian` instead of `RunVaspDirect` for consistency with other FWs. I had previously set custodian via a powerup in `scan.py` but this last commit implements it at the FW level.

rkingsbury commented:
Thanks for the comments @utf, working on them now.
Regarding the tests - the problem I think I have is that the `use_fake_vasp` powerup will pass the same reference dir to BOTH calls to `RunVaspCustodian`, but each call needs a separate set of inputs and outputs. Is there a way to  pass a different ref directory to each call, or to include multiple sets of inputs/outputs in one reference dir?

rkingsbury commented on atomate/vasp/firetasks/write_inputs.py:
```diff
@@ -208,6 +209,34 @@ def run_task(self, fw_spec):
+        vrun = Vasprun("vasprun.xml")
+        bandgap = vrun.get_band_structure().get_band_gap()["energy"]
+        structure = vrun.final_structure
+        vis = MPScanRelaxSet(structure, bandgap=bandgap, **kwargs)
```
@utf I modeled `MPScanRelaxSet` on `MPRelaxSet`, which does not implement a `from_prev_calc` method. How is the information passed in that case? Is there a way I can achieve the desired behavior purely by modifying the FromPrev method in atomate without modifying the inputset?

utf replied:
Good point! It seems like the MPRelaxSet doesn't support inheriting settings from a parent run either.

In that case, it seems unfair to force you to implement this also (although I think it would still be useful in some cases). 

I think this is something we should keep in mind for atomate v2!

utf replied:
For your question. You could implement the FromPrev behavior in atomate but I would definitely advise against it. There should be a separation of input settings (controlled in pymatgen) and writing input sets for calculations (atomate).

rkingsbury commented:
> As a final comment, is there anything special about your preconditioning step? Or, for example, should you be able to run SCAN on top of a standard static calculation? If so, I think you should make the preconditioning step optional (controlled through a variable it the Firework init method).

@utf we've found that running a GGA preconditioning step using the same settings (ENCUT, etc.) significantly helps convergence and run times. In addition to getting the structure closer to optimized, the initial run gives us wavefunctions that can be passed to SCAN. We also use that step to determine the bandgap so that we can set the appropriate k-point density. While it is possible to just run the SCAN by itself, I feel like at that point it's really a different workflow that loses a lot of the advantages we've tried to build in here. It is also possible to use a static instead of a relaxation for the first step, but that could easily be accomplished with setting NSW=0 in `user_incar_settings`.

utf commented:
Ok @rkingsbury that makes sense. Feel free to ignore that comment in that case.

utf commented:
> Regarding the tests - the problem I think I have is that the use_fake_vasp powerup will pass the same reference dir to BOTH calls to RunVaspCustodian, but each call needs a separate set of inputs and outputs. Is there a way to pass a different ref directory to each call, or to include multiple sets of inputs/outputs in one reference dir?

Ah, I see your issue now. It's a shame there isn't a proper unit test for the optimization workflow...

Is there any downside to copying the same folder in both cases?

Your unit test would then check:
- That the input sets are written correctly for both runs.
- That the data is processed correctly after both runs.

That the final calculation results aren't correct shouldn't really matter, as long as the data gets to the database ok...

rkingsbury commented:
> > Regarding the tests - the problem I think I have is that the use_fake_vasp powerup will pass the same reference dir to BOTH calls to RunVaspCustodian, but each call needs a separate set of inputs and outputs. Is there a way to pass a different ref directory to each call, or to include multiple sets of inputs/outputs in one reference dir?
> 
> Ah, I see your issue now. It's a shame there isn't a proper unit test for the optimization workflow...
> 
> Is there any downside to copying the same folder in both cases?
> 
> Your unit test would then check:
> 
>     * That the input sets are written correctly for both runs.
> 
>     * That the data is processed correctly after both runs.
> 
> 
> That the final calculation results aren't correct shouldn't really matter, as long as the data gets to the database ok...

@utf thanks again for all the feedback. Per our discussions I've added unit tests that verify that certain INCAR parameters are being correctly set and unset at different parts of the workflow, and that the KSPACING parameter is being correctly calculated from the bandgap in the first relaxation. I use `RunVaspFake` to copy GGA results (corresponding to the .relax1 step), which means that the final results are incorrect, but that isn't relevant to the test. 

I've included four sets of test files corresponding to materials with small, large, and zero bandgaps to test the bounds on KSPACING, and also one set that includes VdW corrections to verify that the associated INCAR flags are being properly toggled.

I've also renamed `WriteVaspStaticFromPrev` to `UpdateScanRelaxBandgap` as requested.

rkingsbury commented:
It looks like the tests are failing because they are looking for POTCARs that are not present:

`OSError: You do not have the right POTCAR with functional PBE_52 and label Li_sv in your VASP_PSP_DIR`

I don't see this issue in my local testing

utf commented:
I presume this is just a case of using the correct docker image. Are these POTCARs needed for Pymatgen tests (i.e., to write the input sets)?

rkingsbury commented:
> I presume this is just a case of using the correct docker image. Are these POTCARs needed for Pymatgen tests (i.e., to write the input sets)?

You're right, these are not included in the pymatgen test_files. The missing PSPs are

`PBE_52 and label Li_sv`
`PBE_54 and label Li_sv`
`PBE_54 and label Al`

and probably PBE_54 and label H and F as well.

utf commented:
Hi @rkingsbury 

My PR to fix the potcar issue (#362) has been merged into master. I'm sorry about the merge conflicts but once those are sorted these tests should pass.

utf commented:
Thanks!
